ncm-spma: Add backend for apt

jrha opened the pull request:
Add a SPMA backend for configuring debian derived systems using apt,
this should be considered a technology preview.

The code currently manages source lists (repositories) and packages and not much else.

Tested at RAL on Ubuntu 14.04.4 LTS and Linux Mint 17.3 Rosa.

Based on and depends on #716, so it might be better to wait until that is merged before attempting too much of a review.

jouvin commented:
Thanks for this great contribution! This is certainly something we may be interested in...

jrha commented:
I should also add that at at 31.4% the test coverage is not yet where I would like it to be, so I'll be working on that when I have some more time.

jrha commented:
Discovered that you can review a subset of changes, so https://github.com/quattor/configuration-modules-core/pull/821/files/339a6e87c28fde81e3ac7bc593296738925b65fa..HEAD can be used for review until #716 is merged.

ned21 commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+        # We use $f here to make Devel::Cover happy
+        my $f = "$repo_dir/$i.list";
+        $self->verbose("Unlinking outdated repository $f");
+        if (!unlink($f)) {
```
This isn't NoAction safe.

stdweird replied:
use `CAF::Path` `cleanup` method

piojo replied:
@stdweird please provide an example of how this code should look like

ned21 commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    my ($self, $repo_dir, $allowed_repos, $allow_user_pkgs) = @_;
+    $self->debug(5, 'Entered cleanup_old_repos()');
+
+    return 1 if $allow_user_pkgs;
```
I know this is a problem with the yum code too but at some point we need to split this into "allow_user_repos" and "allow_user_pkgs". For new code I would say "allow_user_pkgs" means that only -- manually installed packages are allowed to stay but manually configured repos get deleted.

ned21 commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+
+    if (! -d $repo_dir) {
+        $self->verbose("$repo_dir didn't exist. Creating it");
+        if (!eval{mkpath($repo_dir)} || $@) {
```
What happens under NoAction?

jrha replied:
Bad things I assume, I took this code from the yum backend, which handles NoAction in strange ways, I'll take a look at it again.

stdweird replied:
use `CAF::Path` `directory` method

ned21 commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    my $tree_pkgs = $config->getElement($TREE_PKGS)->getTree();
+    my $tree_component = $config->getElement($TREE_COMPONENT)->getTree();
+    # Convert these crappily-defined fields into real Perl booleans.
+    $tree_component->{run} = $tree_component->{run} eq 'yes';
```
If you intend to start converting these fields in other code to be either booleans or text, you should start that effort here by accepting both.

stdweird replied:
i'd rather not. this should be handled in CCM (via some gettree option), before everyone starts to copy paste something that will be some pretty ugly code. (if you open an issue in CCM, i might come up with something)

ned21 commented on ncm-spma/src/test/perl/apt-cleanup-sources.t:
```diff
@@ -0,0 +1,116 @@
+if ($EUID) {
+  chmod(0500, $REPO_DIR);
+} else {
+  diag("Trying to set immutable bit for root user to create unremovable files. ".
```
Surely any running of unit tests root should be an immediate fail?

stdweird replied:
`CAF::Path` methods are unittested, but there is no way to trigger failed mocked operation (like a mcoked failed file removal). open an issue for that in `maven-tools`; it shouldn't be that hard to add.

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+# ${license-info}
```
use new style like
https://github.com/quattor/configuration-modules-core/pull/808/files#diff-b5f0d02e208de8bb63cc453ae9e11a33L1

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+use warnings;
+use NCM::Component;
+our $EC=LC::Exception::Context->new->will_store_all;
+our @ISA = qw(NCM::Component);
```
`use parent qw(NCM::Component);`

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+use NCM::Component;
+our $EC=LC::Exception::Context->new->will_store_all;
+our @ISA = qw(NCM::Component);
+use EDG::WP4::CCM::Element qw(escape unescape);
```
once https://github.com/quattor/CCM/pull/102 is merged, you should import the escape functions from `CCM::Path` (altough current code will still work)

piojo replied:
And it is merged now, so please rely on `CCM::Path`.

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+use CAF::Process;
+use CAF::FileWriter;
+use CAF::FileEditor;
+use LC::Exception qw(SUCCESS);
```
replace with `use CAF::Object qw(SUCCESS);` (we are phasing out `LC`)

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+
+Readonly::Scalar my $DIR_REPOS => "/etc/apt/sources.list.d";
+Readonly::Scalar my $DIR_PREFERENCES => "/etc/apt/preferences.d";
+Readonly::Scalar my $TEMPLATE_REPOS => "spma/apt-source.tt";
```
move the TT files to an apt subdir (and open an issue for all non-common TT to move to a subdir as well)

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+Readonly::Scalar my $TEMPLATE_PREFERENCES => "spma/apt-preferences.tt";
+Readonly::Scalar my $TREE_REPOS => "/software/repositories";
+Readonly::Scalar my $TREE_PKGS => "/software/packages";
+Readonly::Scalar my $TREE_COMPONENT => "/software/components/${project.artifactId}";
```
wherever this is used, use `$self->prefix()` instead

piojo replied:
Absolutely.  `$self->prefix` is what lets component aliasing work.

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+
+    my $current = Set::Scalar->new(map(m{(.*)\.list$}, readdir($dir)));
+
+    closedir($dir);
```
replace all the above with `my $current = Set::Scalar->new(glob("$repo_dir/*.list"));`
non-existing directories should be handled with `CAF::Path` diretcory method. 
do a `use parent (NCM::Component CAF::Path);` above and `$self->directory_exists($repodir)` as test.

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+
+    closedir($dir);
+    my $allowed = Set::Scalar->new(map($_->{name}, @$allowed_repos));
+    my $rm = $current-$allowed;
```
`$rm` -> `$to_remove`

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    closedir($dir);
+    my $allowed = Set::Scalar->new(map($_->{name}, @$allowed_repos));
+    my $rm = $current-$allowed;
+    foreach my $i (@$rm) {
```
`$i` -> `$repo`?

piojo replied:
Keep locals short, please. `$i` or `$r` are more than enough for this little loop.

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    my $rm = $current-$allowed;
+    foreach my $i (@$rm) {
+        # We use $f here to make Devel::Cover happy
+        my $f = "$repo_dir/$i.list";
```
`$f` -> `$fn`

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    my ($self, $repo_dir) = @_;
+    $self->debug(5, 'Entered initialize_repos_dir()');
+
+    if (! -d $repo_dir) {
```
test with `CAF::Path` `directory_exists` (or not at all, i think `CAF::Path` `directory` doesn't mind existing directories)

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+
+# Returns the URLs of the repositories listed in $prots, with their
+# host part replaced by each of the reverse @proxies.
+sub generate_reverse_proxy_urls
```
is this code similar to other code? like yum or aii? (you can use existing methods in other modules :smile: )

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    my ($self, $prots, @proxies) = @_;
+    $self->debug(5, 'Entered generate_reverse_proxy_urls()');
+
+    my @l;
```
no single letter variables

piojo replied:
Locals should be as short as possible. :)

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    foreach my $pt (@$prots) {
+        foreach my $px (@proxies) {
+            my $url = $pt->{url};
+            $url =~ s{^(.*?):(/+)[^/]+}{$1:$2$px};
```
should you handle the case that this regex doesn't match anything? it also looks like a confusiong way to write 
```perl
push(@l, ($pt->{url} =~ m/^(.*?):(/+)/) ? "$1:$2$px" : $pt->{url})
```

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+
+
+# Generate the URLs for the forward proxies, based on whether the
+# repository is accessible over HTTP or HTTPS.
```
i fail to see how the code below has anythig to do with `https?`

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+
+
+# Generates the repository files in $repos_dir based on the contents
+# of the $repos subtree. It uses Template::Toolkit $template to render
```
it should use `CCM::TextRender`, the `template` moethdo was removed in https://github.com/quattor/ncm-ncd/pull/68
see https://github.com/quattor/configuration-modules-core/pull/808 for an example

piojo replied:
Please provide an example here to help the developer. :smile:

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    foreach my $repo (@$repos) {
+        my $prots = $repo->{protocols}->[0];
+        my $fh = CAF::FileWriter->new("$repos_dir/$repo->{name}.list", log => $self);
+        print $fh "# File generated by ", __PACKAGE__, ". Do not edit\n";
```
pass this via a header to textrender, or in the TT file, or not at all

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+                $repo->{proxy} = $self->generate_forward_proxy_urls($repo->{protocols}, @px);
+            }
+        }
+        my $rs = $self->template()->process($template, $repo, $fh);
```
use `CCM::TextRender`

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    my $self = shift;
+    $self->debug(5, 'Entered installed_pkgs()');
+
+    my $cmd = CAF::Process->new($CMD_DPKG_QUERY, keeps_state => 1, log => $self);
```
as you do not use `$cmd` below, do a `my $out = ...->output();`

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+
+    my $out = $cmd->output();
+    if ($?) {
+        return 0;
```
better to return with `return;`, unless the `0` has any significance

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    if ($?) {
+        return 0;
+    }
+    my @pkgs = grep(m{^ii }, split(/\n/, $out));
```
only a ` ` or `\s`?

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+        return 0;
+    }
+    my @pkgs = grep(m{^ii }, split(/\n/, $out));
+    for (@pkgs) {s/^...;//g}; # Strip package status characters
```
? (esp the `/g` in combination with the `^`)
 add it with a map above 
```perl
my @pkgs = map {s/^.{3};//} grep {m/^ii\s/} split("\n", $out);
```

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+}
+
+
+sub get_package_version_arch
```
small comment/docstring

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+            }
+        }
+    }
+    else {
```
use `} else {` on one line

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+                    $self->debug(5, '  Adding package ', $name, ' with version ', $ver, ' and architecture ', $arch, ' to list');
+                    push(@versions, sprintf('%s:%s=%s', $name, $arch, unescape($ver)));
+                }
+            }
```
one line

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    my @versions;
+
+    if ($details) {
+        while (my ($ver, $archs) = each(%$details)) {
```
each is evil https://github.com/quattor/CAF/issues/109 (and not sorted)

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+}
+
+
+sub apply_package_version_arch
```
docstring / comment

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    my @packages = @$packagelist;
+    my @results;
+
+    while (my $name = shift(@packages)) {
```
simple loop thorugh them, the `while/shift` adds nothing
```perl
foreach my $name (@$packagelist) {
```

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    while (my $name = shift(@packages)) {
+        my $name_escaped = escape($name);
+        if (exists($packagetree->{$name_escaped})) {
+            my $versions = $self->get_package_version_arch($name, $packagetree->{$name_escaped});
```
does the `apt` support globs in package name like `kernel*`

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+            push(@results, @$versions);
+        }
+        else {
+            $self->error('Could not find package ', $name, ' in tree');
```
i would collect all packages that are not found (`push*(@notfound, $name)`) , and once report one error
```perl
$self->error("Could not find packages", join(", ", @notfound)) if @notfound`
```

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+        my $name = unescape($name_escaped);
+        if (!$name) {
+            $self->error("Invalid package name: ", unescape($name));
+            return 0;
```
return in the middle of an each... :see_no_evil: :hear_no_evil: :speak_no_evil:  https://github.com/quattor/CAF/issues/109

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+
+    while (my $name_escaped = each(%$pkgs)) {
+        my $name = unescape($name_escaped);
+        if (!$name) {
```
any examples?

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    $self->debug(5, 'Entered resynchronize_package_index()');
+
+    my $cmd = CAF::Process->new($CMD_APT_UPDATE, keeps_state => 1, log => $self);
+    $cmd->execute() or return 0;
```
```perl
return $cmd->execute() ? 1 : undef;
```

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    my ($self) = @_;
+    $self->debug(5, 'Entered upgrade_packages()');
+
+    my $cmd = CAF::Process->new($CMD_APT_UPGRADE, keeps_state => 1, log => $self);
```
this is not `keeps_state => 1`

piojo replied:
Indeed. Please remember that `keeps_state => 1` makes the command be executed even under `NoAction`. It is meant for trivial things like `ls`, `ip addr show` and the like.

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    my ($self, $packages) = @_;
+    $self->debug(5, 'Entered install_packages()');
+
+    my $cmd = CAF::Process->new([@$CMD_APT_INSTALL, @$packages], keeps_state => 1, log => $self);
```
also not `keeps_state => 1`

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    my ($self, $packages) = @_;
+    $self->debug(5, 'Entered mark_packages_auto()');
+
+    my $cmd = CAF::Process->new([@$CMD_APT_MARK, 'auto', @$packages], keeps_state => 1, log => $self);
```
is this `keeps_state => 1`? i think not.

piojo replied:
Also, line too long.

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    my ($self, $packages) = @_;
+    $self->debug(5, 'Entered autoremove_packages()');
+
+    my $cmd = CAF::Process->new([@$CMD_APT_AUTOREMOVE, @$packages], keeps_state => 1, log => $self);
```
no keeps_state

piojo replied:
Line too long, but that is fixed as soon as `keeps_state` is removed.

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    $self->debug(5, 'Entered Configure()');
+
+    # Get configuration trees
+    my $tree_repos = $config->getElement($TREE_REPOS)->getTree();
```
new style
```perl
$config->getTree($TREE_REPOS);
```

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    my $tree_component = $config->getElement($TREE_COMPONENT)->getTree();
+    # Convert these crappily-defined fields into real Perl booleans.
+    $tree_component->{run} = $tree_component->{run} eq 'yes';
+    $tree_component->{userpkgs} = defined($tree_component->{userpkgs}) && $tree_component->{userpkgs} eq 'yes';
```
make it mandatory in the schema with default false?

stdweird replied:
or at least `$tree_component->{userpkgs} = ($tree_component->{userpkgs} || "no") eq 'yes';`

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    $tree_component->{run} = $tree_component->{run} eq 'yes';
+    $tree_component->{userpkgs} = defined($tree_component->{userpkgs}) && $tree_component->{userpkgs} eq 'yes';
+
+    $self->initialize_repos_dir($DIR_REPOS);
```
check return codes

piojo replied:
Yup. We are not in a position to accept exceptions yet.

```perl
$self->initialize_repos_dir($DIR_REPOS) or return 0;
```

and so on.

stdweird commented on ncm-spma/src/main/resources/apt-source.tt:
```diff
@@ -0,0 +1,17 @@
+# Owner: [% owner %]
+
+[%- IF gpgcheck -%]
+[%- SET trust = "trusted=yes " -%]
```
indent

stdweird replied:
or better yet
```
[% trust = gpgcheck ? "trusted=yes " : "" %]
```

stdweird replied:
or below, with the inermediate `trust`

stdweird commented on ncm-spma/src/test/perl/apt-cleanup-sources.t:
```diff
@@ -0,0 +1,116 @@
+
+Readonly my $REPO_DIR => "target/test/clean.sources.list.d";
+
+sub initialize_repos
```
this is not needed if you switch to `CAF::Path`, it's all mocked

stdweird commented on ncm-spma/src/test/perl/apt-list-installed.t:
```diff
@@ -0,0 +1,40 @@
+use Test::More;
+use NCM::Component::spma::apt;
+
+plan skip_all => "Cannot test get_installed_pkgs, dpkg not present" if ! -x "/usr/bin/dpkg";
```
? why is this needed? should be mocked, if not, open issue in `maven-tools` for whatever is missing.

stdweird commented on ncm-spma/src/test/perl/apt-sources-generate.t:
```diff
@@ -0,0 +1,253 @@
+ok(defined($fh), "Correct file opened");
+
+my $name = $repos->[0]->{name};
+like("$fh", qr{^\# Name: $name$}m, "Source got the correct name");
```
use TT unittests rather than reinventing these here.

piojo replied:
Provide an example, please :)

piojo commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+use Text::Glob qw(match_glob);
+use Readonly;
+
+Readonly::Scalar my $DIR_REPOS => "/etc/apt/sources.list.d";
```
```perl
Readonly my $DIR_REPOS => "/etc/apt/sources.list.d";
```
is shorter and more readable IMHO. Not a big deal, though.

piojo commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+    my $current = Set::Scalar->new(map(m{(.*)\.list$}, readdir($dir)));
+
+    closedir($dir);
+    my $allowed = Set::Scalar->new(map($_->{name}, @$allowed_repos));
```
I know standard Quattor code relies on parentheses a lot. But Perl usually prefers

```perl
map {$_->{name}}, @$allowed_repos
```
Parentheses in `map`, `grep` and `sort` cause lots of headaches because of ambiguities.

piojo commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,433 @@
+
+    my $packages_installed = $self->get_installed_pkgs();
+    my $packages_desired = $self->get_desired_pkgs($tree_pkgs);
+    my $packages_unwanted = $packages_installed->difference($packages_desired);
```
Sets support direct substraction, which I find nicer:

```perl
my $packages_unwanted = $packages_installed - $packages_desired;
```

jrha commented:
I managed to devote some time to this today, it's looking much better but not quite there yet... still needs more work.

stdweird commented on ncm-spma/src/main/pan/components/spma/apt/schema.pan:
```diff
@@ -0,0 +1,16 @@
+type component_spma_apt = {
+    include structure_component
+    include component_spma_common
+    @{ Allow user defined (i.e. unmanaged) repositories to be present on the system }
```
ah, the shortest way to for users to take over the machine. you sure you want this?

jrha replied:
Yes!
It's something @ned21 suggested, currently `userpkgs` implies this. Having a separate switch for repositories which defaults to `false` seems like an improvement.

stdweird replied:
hmm, true. never tought about it. do we allow userpkgs = false with userrepos = true?

jrha replied:
Why not, it would make sense if you wanted something else (e.g. RedHat Satellite) managing your repositories.

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+
+=head1 NAME
+
+NCM::Component::spma::apt - NCM SPMA backend for apt
```
all modules `C<>`

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+
+=head1 DESCRIPTION
+
+apt.pm implements an apt backend for ncm-spma, the approach taken is to defer as much work as possible to apt.
```
`apt.pm` -> `This plugin`?

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+
+=item
+
+Setup source directory if required
```
out of interest, does the pod look different if you put this after the `=item` or in a new paragraph like this?

jrha replied:
Nope! I just tried:

```perl
=item Remove sources that are not found in the profile

=item

Update source cache from upstream sources
```

For which `pod2html` generates:

```html
<li><p>Remove sources that are not found in the profile</p>

</li>
<li><p>Update source cache from upstream sources</p>

</li>
```

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+
+=back
+
+Support for groups (tasks) is not currently provided.
```
i tought we were going to remove it?

jrha replied:
Good point.

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+Only a very minimal schema is implemented.
+
+Sources listed under C</software/repositories> will be configured,
+URLs should be followed by the suite and sections required e.g. http://example.org/debian unstable main
```
but the example in `C<>`

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+use Text::Glob qw(match_glob);
+use Readonly;
+
+Readonly::Scalar my $DIR_SOURCES => "/etc/apt/sources.list.d";
```
you can  drop the `::Scalar`

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+    my $allowed = Set::Scalar->new(map($_->{name}, @$allowed_sources));
+    my $to_remove = $current-$allowed;
+
+    if (-d $sources_dir) {
```
use `CAF::Path` and `->directory_exists`

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+        foreach my $s (@$to_remove) {
+            my $fn = "$sources_dir/$s.list";
+            $self->verbose("Unlinking outdated source $fn");
+            if (!defined (CAF::Path->cleanup($fn, ""))) {
```
with `use CAF::Path` i meant, add `CAF::Path` to the parent list at the end do `$self->cleanup()`. this probably doesn't work like you think it does.

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+    my ($self, $sources_dir) = @_;
+    $self->debug(5, 'Entered initialize_sources_dir()');
+
+    if (CAF::Path->directory($sources_dir) != $sources_dir) {
```
this is not what you want. for string comparison you need `ne`, but on failure, this will give a warning, since `->directory` retruns undef.
you can do
```perl
if (! $self->directory($sources_dir)) {
    $self->error("Unable to create source dir $sources_dir: $self->{fail}");
    return 0;
}
```

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+    return Set::Scalar->new(@pkgs);
+}
+
+# Build list of packages that should be present on system from configuration tree
```
is this correct? what do `name` and `details` do?

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+    my @versions;
+
+    if ($details) {
+        foreach my $version (keys %$details) {
```
`sort keys`

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+            my $params = $details->{$version};
+            if ($params->{arch}) {
+                foreach my $arch (keys %{ $params->{arch} }) {
+                    $self->debug(5, '  Adding package ', $name, ' with version ', $version, ' and architecture ', $arch, ' to list');
```
unescape version?

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+                    push(@versions, sprintf('%s:%s=%s', $name, $arch, unescape($version)));
+                }
+            } else {
+                $self->debug(5, '  Adding package ', $name, ' with version ', $version, ' but without architecture to list');
```
unescape version?

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+        foreach my $version (keys %$details) {
+            my $params = $details->{$version};
+            if ($params->{arch}) {
+                foreach my $arch (keys %{ $params->{arch} }) {
```
sort keys

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+        } else {
+            push(@notfound, $name);
+        }
+        $self->error("Could not find packages", join(", ", @notfound)) if @notfound;
```
do this outside the loop?

stdweird replied:
also add a space after `packages` (all arguments to reporter methods are appended, no spacing is added)

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+
+    my $packages = Set::Scalar->new();
+
+    foreach my $name_escaped (keys %$pkgs) {
```
sort keys

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+    foreach my $name_escaped (keys %$pkgs) {
+        my $name = unescape($name_escaped);
+        if (!$name) {
+            $self->error("Invalid package name: ", unescape($name));
```
just `$name`? otherwsie double unescaped name

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+
+    foreach my $name_escaped (keys %$pkgs) {
+        my $name = unescape($name_escaped);
+        if (!$name) {
```
this is an odd test. why is it here?

jrha replied:
What does unescape return if it fails to unescape?

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+    my $self = shift;
+    $self->debug(5, 'Entered resynchronize_package_index()');
+
+    my $cmd = CAF::Process->new($CMD_APT_UPDATE, log => $self);
```
if it is only updating package metadata, you could do `keeps_state => 1`; if it does more than that, please update the comment

jrha replied:
Isn't updating the system's metadata cache something that changes the state of the system?

ned21 replied:
I would "yes" but it can be difficult to make NoAction give useful info if you don't have an up to date cache so I think keeps_state=>1 might be OK?

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+    $self->debug(5, 'Entered upgrade_packages()');
+
+    my $cmd = CAF::Process->new($CMD_APT_UPGRADE, log => $self);
+    $cmd->execute() or return 0;
```
do a `return $cmd->execute() ? 1 : 0` as in previous method

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+    $self->debug(5, 'Entered install_packages()');
+
+    my $cmd = CAF::Process->new([@$CMD_APT_INSTALL, @$packages], log => $self);
+    $cmd->execute() or return 0;
```
same as above

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+    $self->debug(5, 'Entered mark_packages_auto()');
+
+    my $cmd = CAF::Process->new([@$CMD_APT_MARK, 'auto', @$packages], log => $self);
+    $cmd->execute() or return 0;
```
same as above

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+    $self->debug(5, 'Entered autoremove_packages()');
+
+    my $cmd = CAF::Process->new([@$CMD_APT_AUTOREMOVE, @$packages], log => $self);
+    $cmd->execute() or return 0;
```
same as above

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+
+    # Get configuration trees
+    my $tree_sources = $config->getTree($TREE_SOURCES);
+    my $tree_pkgs = $config->getElement($TREE_PKGS)->getTree();
```
no `getElement`, use getTree on config directlyt as previous line

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,385 @@
+    # Get configuration trees
+    my $tree_sources = $config->getTree($TREE_SOURCES);
+    my $tree_pkgs = $config->getElement($TREE_PKGS)->getTree();
+    my $tree_component = $config->getElement($self->prefix())->getTree();
```
same remark

stdweird commented on ncm-spma/src/test/perl/apt-apply-package-version-arch.t:
```diff
@@ -0,0 +1,52 @@
+
+my $cmp = NCM::Component::spma::apt->new("spma");
+
+cmp_bag($cmp->apply_package_version_arch($set, $tree), $desired_details, "Package details translated");
```
why not `is_deeply` from `Test::More`?

jrha replied:
`is_deeply` will complain about result are ordering, `cmp_bag` will not.

stdweird replied:
ok, if the order is not relevant.

stdweird commented on ncm-spma/src/test/perl/apt-cleanup-sources.t:
```diff
@@ -0,0 +1,96 @@
+
+# Restore permissions on the source for future executions.
+if ($EUID) {
+  chmod(0700, $SOURCE_DIR);
```
fix indent

stdweird commented on ncm-spma/src/test/perl/apt-get-package-version-arch.t:
```diff
@@ -0,0 +1,42 @@
+
+my $cmp = NCM::Component::spma::apt->new("spma");
+
+cmp_bag($cmp->get_package_version_arch('example-package', $details), $desired_details, "Package details translated");
```
`is_deeply`

stdweird commented on ncm-spma/src/test/perl/apt-list-installed.t:
```diff
@@ -0,0 +1,38 @@
+
+my $cmp = NCM::Component::spma::apt->new("spma");
+
+my $pkgs = $cmp->get_installed_pkgs();
```
how does this work without mocked command outpput?

jrha replied:
It doesn't?

stdweird commented on ncm-spma/src/test/perl/apt-sources-generate.t:
```diff
@@ -0,0 +1,111 @@
+
+my $mock = Test::MockModule->new('CAF::FileWriter');
+
+$mock->mock('cancel', sub {
```
why is this needed?

stdweird commented:
@jrha nice progress!

ned21 commented on ncm-spma/src/main/pan/components/spma/yum/schema.pan:
```diff
@@ -65,6 +65,10 @@ type component_spma_yum = {
     "plugins" ? spma_yum_plugins
     "main_options" ? spma_yum_main_options
+    @{ Run the SPMA after configuring it }
+    "run" ? legacy_binary_affirmation_string
```
I don't believe the yum backend supports this attribute so it should not appear here?

jrha replied:
Good point, removed.

ned21 commented on ncm-spma/src/main/pan/components/spma/yumng/schema.pan:
```diff
@@ -66,6 +66,10 @@ type component_spma_yum = {
+    "userpkgs" ? legacy_binary_affirmation_string
 };
 
 bind "/software/components/spma" = component_spma_yum;
```
Should be component_spma_yumng instead?

jrha replied:
I've been working on cleaning up the schema in a seperate branch, I think this would be better fixed there.

jrha commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,386 @@
+    my $allowed = Set::Scalar->new(map($_->{name}, @$allowed_sources));
+    my $to_remove = $current-$allowed;
+
+    if ($self->directory_exists($sources_dir)) {
```
@stdweird I must be having another stupid moment - this is giving me problems with the unit tests.
`directory_exists` appears to get mocked by `Test:Quattor` and then doesn't behave as expected, replacing the line with `if (-d $sources_dir) {` allows the tests to pass but that just seems like a bad idea.

stdweird replied:
what test is failing? you also need latest build-tools and everything has to be mocked, but should all work

jrha replied:
`apt-cleanup-sources.t` is the source of my pain, coming from the yum backend it's mostly not mocked and I'm getting confused.

jrha replied:
:relieved:  I've rewritten the whole set of tests with lots of mocking.

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,394 @@
+
+
+# Wrap glob builtin to allow for unittesting
+sub _glob
```
use https://github.com/quattor/CAF/pull/230, which has support for Test::Quattor unittests

stdweird replied:
@jrha ping?

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,394 @@
+    my ($self, $sources_dir, $sources, $template, $proxy, $type, $port) = @_;
+    $self->debug(5, 'Entered generate_sources()');
+
+    #my @px = split(",", $proxy) if $proxy;
```
add an issue to fix/address or remove the comments

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,394 @@
+    foreach my $source (@$sources) {
+        $self->debug(5, "Generating source: $source->{name}");
+        my $prots = $source->{protocols}->[0];
+        # if (!exists($source->{proxy}) && @px) {
```
open an issue to include them or remove these comments

stdweird commented on ncm-spma/src/test/perl/apt-cleanup-sources.t:
```diff
@@ -0,0 +1,117 @@
+my $sourcefiles;
+
+
+$mock->mock('_glob', sub {
```
if you use listdir from CAF::Path, this is not needed

stdweird commented on ncm-spma/src/test/perl/apt-cleanup-sources.t:
```diff
@@ -0,0 +1,117 @@
+});
+
+
+$mock->mock('cleanup', sub {
```
whole of CAF::Path is already mocked and integrated in Test::Quattor 1.54 (even the not yet merged CAF::Path listdir)

stdweird commented on ncm-spma/src/test/perl/apt-cleanup-sources.t:
```diff
@@ -0,0 +1,117 @@
+});
+
+
+$mock->mock('directory_exists', sub {
```
no need to mock this. you can create a mock directory within Test::Quattor by either creating a mocked file (which will create the mocked parent tree) or directly creating a mocked directory.
example below or in PR, lets see what is easier

stdweird commented on ncm-spma/src/test/perl/apt-cleanup-sources.t:
```diff
@@ -0,0 +1,117 @@
+
+=cut
+
+$immutable = 0;
```
remove these lines and mocking above, just delete the directory using either

CAF::Path cleanup

```perl
$cmp->cleanup($SOURCE_DIR);
```
or Test::Quattor function
```perl
remove_any($SOURCE_DIR);
```

and that's it. remember, `CAF::Path` is mocked entirely in `Test::Quattor`. the only thing you need to do is to make the fake file/directory structure.

stdweird commented on ncm-spma/src/test/perl/apt-cleanup-sources.t:
```diff
@@ -0,0 +1,117 @@
+
+=cut
+
+$immutable = 0;
```
remove these, just create the files, parent dir will exist.

```perl
foreach my $fn (qw(source1 source2)) {
  set_file_contents("$SOURCE_DIR/$_.list", "")
};
```
this will create the parent dir and "touch" the files.

stdweird commented on ncm-spma/src/test/perl/apt-cleanup-sources.t:
```diff
@@ -0,0 +1,117 @@
+$sourcefiles = Set::Scalar->new(map { "$SOURCE_DIR/$_.list" } qw(source1 source2));
+
+ok($cmp->cleanup_old_sources($SOURCE_DIR, $sources), "Cleanup old source files");
+ok($sourcefiles->contains("$SOURCE_DIR/source1.list"), "Preserve allowed source file");
```
this should then become 

```
ok($cmp->file_exists("$SOURCE_DIR/source1.list"), "Preserve allowed source file");
ok(!$cmp->file_exists("$SOURCE_DIR/source2.list"), "Remove unwanted source file");
```

stdweird commented on ncm-spma/src/test/perl/apt-cleanup-sources.t:
```diff
@@ -0,0 +1,117 @@
+$directory_exists = 1;
+$sourcefiles = Set::Scalar->new(map { "$SOURCE_DIR/$_.list" } qw(source1 source2));
+
+ok(!$cmp->cleanup_old_sources($SOURCE_DIR, $sources), "Report error when an outdated repo cannot be removed");
```
i guess you want to test a failed cleanup? this is something you can't do yet. i would open an issue to add this later, i plan to add this to maven-tools soonish

stdweird commented on ncm-spma/src/test/perl/apt-sources-generate.t:
```diff
@@ -0,0 +1,108 @@
+use CAF::FileWriter;
+use Cwd;
+
+$CAF::Object::NoAction = 1;
```
if you use `Test::Quattor`, this is only needed to test noaction behaviour, not to prevent that the unittests actually write stuff.

stdweird commented on ncm-spma/src/test/perl/apt-sources-generate.t:
```diff
@@ -0,0 +1,108 @@
+
+is($cmp->generate_sources($SOURCES_DIR, $sources, "an invalid template name"), 0, "Invalid template name is detected");
+is($cmp->{ERROR}, 1, "Errors on template rendering are reported");
+$fh = get_file("$SOURCES_DIR/$name.list");
```
you don't do anything with `$fh`

jrha commented:
Tests are failing due to ncm-path.

jrha commented:
@hpcugentbot test this please

stdweird commented:
@jrha conflicts due to common yum schema PR merged

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,379 @@
+
+=cut
+
+use strict;
```
this and next line are part of the header template.

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,379 @@
+        my $prots = $source->{protocols}->[0];
+        my $rs = EDG::WP4::CCM::TextRender->new($template, $source, relpath => 'spma', log => $self);
+        if ($rs) {
+            my $fh = $rs->filewriter("$sources_dir/$source->{name}.list", log => $self);
```
`log => $self` is default

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,379 @@
+
+    my $tr = EDG::WP4::CCM::TextRender->new($TEMPLATE_CONFIG, $config, relpath => 'spma', log => $self);
+    if ($tr) {
+        my $fh = $tr->filewriter($FILE_CONFIG, log => $self);
```
`log => $self` is default

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,379 @@
+}
+
+# For a given package name, extract version and architecture from tree passed in details
+# returns a list of packages formatted with name, version and architecture for use with apt
```
`list` ->  arrayref`

stdweird commented on ncm-spma/src/main/perl/spma/apt.pm:
```diff
@@ -0,0 +1,379 @@
+    return \@versions;
+}
+
+# Apply versions from configuration tree to a bare list of packages
```
what does it return?

jrha commented:
@stdweird I'll do my best to address your comments, but at this point I feel that this PR is better off being merged than not, it may not be perfect, but is probably good enough. On Thursday it will have been open for a year, which makes me sad.

stdweird commented:
@jrha fair enough. expect a PR to your PR soonish 😄

jrha commented:
I'll probably have some fixes in a few minutes, but might not find time to do anything about the test mocking.

stdweird commented:
ok, i can certainly take care of the test remarks. lets try to merge this by thursday.

jrha commented:
@stdweird I've addressed all the non-test stuff I can, rebased and generally tidied up, if you want to sort out the tests to your satisfaction please go ahead!

stdweird commented:
@jrha still thursday somewhere 😄

jrha commented:
It is in Tahiti!
